firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,104 @@
+/* compute_body_crc
+Scenario: CRCElement: The compute_body_crc function correctly computes crc from body
+  GIVEN("A CRC element initialized with an empty payload")
+    WHEN("the crc is computed from the payload")
+      THEN("the computed crc is equal to the expected crc")
+
+  GIVEN("A CRC Element initialized with the payload '123456789'")
+    WHEN("the crc is computed from the payload")
+      THEN("the computed crc is equal to the expected crc")
+
+  GIVEN("A crc element is initalised with payload 0x00")
+    WHEN("the crc is computed from the payload")
+      THEN("the computed crc is equal to the expected crc")
+
+  GIVEN("A crc element is initalised with payload 0x01")
+    WHEN("the crc is computed from the payload")
+      THEN("the computed crc is equal to the expected crc")
+
+/* write function
+Scenario: CRCElement: The write function correctly generates body from computed crc and payload
+  GIVEN("A crc element initalised with payload=[0x01, 0x02, 0x05]")
+    WHEN("The crc and payload are written to the output buffer")
+      THEN("crc computed in write method is same as crcchecker compute")
+      THEN("The output buffer is as expected")
+
+  GIVEN("A crc element is initalised with an empty payload")
+    WHEN("The crc and payload are written to the output buffer")
+      THEN("The CRC returned by the crc() method is equal to 0")
+      THEN("The output buffer is as expected")
+
+/* parse function
+Scenario: CRCElement: The parse function correctly updates internal crc and payload fields from the input buffer
+
+  GIVEN("A CRC Element initialized with an empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc more than 4 bytes and equal to the crc of the payload is parsed")
+      THEN("the parse method reports ok status")
+      THEN("crc returned from the crc() method is same as the crc of the payload in input body")
+      THEN("payload returned from the payload() method is equal to the payload from the body")
+
+    WHEN("A body of less than 4 bytes is parsed")
+      THEN("the parse method reports out of bounds status")
+
+    WHEN("A body with crc and payload is given, where the crc is not equal to the actual crc of the payload")
+      THEN("the parse method reports ok status")
+      THEN("The crc returned from the crc() method is equal to the crc in the body/input_buffer")
+      THEN("The crc returned from the crc() method is not equal to the actual crc for the given payload")
+      THEN("The payload returned from the payload() method is equal to the payload from the body")
+
+/* write parse roundtrip
+Scenario: The crc element correctly writes and parses to a buffer
+
+  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc and payload is generated, parsed and generated again")
+      # write  
+      THEN("The status of write function returns ok")
```

Comment:
```suggestion
      THEN("The first write method call returns ok status")
```
We should specify first vs. second for the THENs asserting things about the write method calls.